design-proposals/klip-36-grace-stream-stream-joins.md

```diff
@@ -0,0 +1,102 @@
+the grace period for stream-stream joins allows more control over the disk space utilization.
+This is because a window needs to stay on disk until the grace period has elapsed in order to
+ensure that any events that come in timestamped during an old window can be processed as long
+as the grace period is active. This means that for windows dramatically smaller than the grace
+period (e.g. a minute) we need to keep that level of granularity for the default 24 hour window.
+
+## What is in scope
+
+- Syntax for specifying grace period within stream-stream joins
+
+## What is not in scope
+
+- N/A
+
+## Value/Return
+
+Enables onboarding use cases that perform stream-stream joins on high cardinality, high throughput
+datasets with small windows without exhausting too much disk space.
+
+## Public APIS
+
+Stream-stream joins require use of `WITHIN` clauses:
+
+```
+JOIN <stream> WITHIN <time unit> ON <condition>
+```
+
+This will be expanded to support a more complex `withinExpression` that leverages the same
+syntax as aggregation windows:
+
+```
+JOIN <stream> WITHIN (SIZE <time unit>, GRACE PERIOD <time unit>) ON <condition>
+```
+
+The old syntax will still be supported for backwards compatibility.
+
+## Design
+
+The `WITHIN` clause will be converted into a `KsqlWindowExpression`, and then converted into
+a `JoinWindows grace(final Duration afterWindowEnd)` call on the `JoinWindows` in the 
+`StreamStreamJoinBuilder`.
+
+## Test plan
+
+We will add the usual QTT tests to ensure that the system respects the new retention limits
+and manually test to ensure that we clean up the RocksDB state for expired windows.
+
+## LOEs and Delivery Milestones
+
+Small LOE - under 2 weeks of engineering time to deliver the syntax and implementation.
+
+## Documentation Updates
+
+The entry in `join-streams-and-tables.md` will be updated to include the following:
+
+```md
+When you join two streams, you must specify a WITHIN clause for matching
+records that both occur within a specified time interval and optionally a
+grace period. The WITHIN clause will specify the "look back" period on 
+the non-triggering stream while the grace period will define how late a record
+may arrive before being processed.
+
+For valid time units, see [Time Units](../syntax-reference.md#time-units).
+
+Here's an example stream-stream-stream join that combines `orders`, `payments` and `shipments` 
+streams. The resulting ``shipped_orders`` stream contains all orders paid within 1 hour of when
+the order was placed, and shipped within 2 hours of the payment being received. 
+
+   CREATE STREAM shipped_orders AS
+     SELECT 
+        o.id as orderId 
+        o.itemid as itemId,
+        s.id as shipmentId,
+        p.id as paymentId
+     FROM orders o
+        INNER JOIN payments p WITHIN 1 HOURS ON p.id = o.id
+        INNER JOIN shipments s WITHIN (SIZE 2 HOURS, GRACE PERIOD 30 MINUTES) ON s.id = o.id;
```

Comment:
This seems like some optimization potential? If one joins specifies a grace-period, we could auto-set it for the other join, too, saving storage but not losing any results?

Reply:
Shouldn't it be valid to have different grace periods on different windows? Why would we want to auto-set it for the other window? I think you understand something I don't 😂

Reply:
Well, you can of course allow this, but it won't buy the user anything (depending on the order of the joins). It seems the lower grace period of the second join limits what you get in your result anyway? If the first join allows to join two records that are 20h old, the result timestamp of the intermediate record would be 20h, too, and thus the second join would drop it.

There would be some details to figure out, ie, maybe think hard (or formally prove) that it's safe to reduce the grace period (but this might not necessarily be part of the KIP I guess -- was just a thought and we might want to track it at least with a ticket).

Reply:
I figured your reasoning would be something along those lines, but I wasn't sure that that hold for all window size/grace period combinations. I need to wrap my head around it a little more (I'll try to get back with a more concrete example), but I'm not convinced that that's always safe to drop - e.g. what would happen if the second window size was much larger than the first window size in your example?

Reply:
Agreed. It's not totally straightforward when we can do what... And as I said, we don't need to make it part of the KLIP. But we should have a ticket "Optimize grace periods for chained stream-stream joins" (maybe with one or two illustrating examples) -- just to make sure we get to it at some point. We also don't need to implement it as part of the KLIP.

Reply:
I am a bit confused here by terminology. The `WITHIN` clause applies to the non-triggering join argument/to the join argument on which it is specified? This means, in this example that we check the timestamp of the record of payments? And the `GRACE` clause applies as well to the join argument on which it is specified? In this example, shipments? So, to accept out-of-order records we just check the timestamp of shipment records?

Reply:
The grace period is base on "stream-time" progress. Ie, if an out-of-order record comes in, with `ts < stream-time - grace-period` the record is dropped. So records from either input may be dropped as late. Note that "stream-time" is advanced based on records of _both_ input streams.

Does this answer your question?